### PR TITLE
Config and script updates

### PR DESCRIPTION
**Pull request by scottcanoe**
This update updates configs and adds several files to the `scripts` directory for plotting and analysis.

Config updates:
 - `pretraining_experiments.py`: Added `10distinctobject` variants for the single-LM distant agent, surface agent, and touch agent experiments. These are primarily for use with new multimodal experiments.
 - `fig4_rapid_inference_with_voting.py`: Updated `half_lms_match` and `fixed_min_lms_match` configs.
 - `fig5_rapid_inference_with_model_based_policies.py`: Added surface agent experiments with and without hypothesis-driven actions.
 - `fig7_flops.py`: Updated logging settings to prevent overwriting of results.
 - `fig8_multimodal_transfer.py`: Switched to using `10distinctobject` configs.
 - `fig9_structured_object_representations.py`: Changed logging settings to record evidence counts. There's a big addition of a custom `MontyExperiment` class to get selective logging working. The change here should eventually go into `tbp.monty`, but it's defined here right now.
 - `visualizations.py` was added. This is where experiments will go used just for generating special plots.
 
New scripts:
 - Adapted and moved some scripts over to make overview plots. Useful modules are `data_utils.py` and `plot_utils.py`. See `overviews.py` to make basic summary plots.

Other changes:
 - `DMC_ROOT_DIR` and `DMC_ANALYSIS_DIR` are now environment variables.
 - READMEs have been updated.
 - Many small changes to imports to keep up with changes to where `tbp.monty` keeps habitat-related configs.

### CODE REVIEW

File: dmc/configs/fig8_multimodal_transfer.py [hunk 1]
```diff
@@ -93,48 +98,62 @@
     ),
 )
 
-# - Add noise and use 5 random rotations.
-touch_agent_1lm_randrot_noise = make_randrot_noise_variant(touch_agent_1lm)
-
+# Distant agent
+dist_agent_1lm_randrot_noise_10distinctobj = deepcopy(dist_agent_1lm_randrot_noise)
```
Comment (nielsleadholm) [inline dmc/configs/fig8_multimodal_transfer.py:102]: Will this be using the correct pre-trained model? As far as I can see `model_name_or_path` isn't updated?

File: dmc/configs/fig9_structured_object_representations.py [hunk 2]
```diff
@@ -20,28 +20,201 @@
...
+)
+dist_agent_1lm_randrot_noise_10simobj["logging_config"] = (
+    DetailedEvidenceLMLoggingConfig(
+        output_dir=str(DMC_RESULTS_DIR),
+        run_name="dist_agent_1lm_randrot_noise_10simobj",
+        wandb_group="dmc",
+        monty_log_level="SELECTIVE",
+    )
 )
+# dist_agent_1lm_randrot_noise_10simobj["logging_config"].monty_handlers.append(
```
Comment (nielsleadholm) [inline dmc/configs/fig9_structured_object_representations.py:215]: Maybe worth removing this to avoid confusion.

**Review by nielsleadholm — COMMENTED** [review]
Really nice, thanks for making all these updates Scott!

Just a few small comments/questions.

File: dmc/configs/fig9_structured_object_representations.py [hunk 3]
```diff
@@ -20,28 +20,201 @@
...
+could be applied to tbp.monty.
+"""
+
+
+class EvidenceLoggingMontyObjectRecognitionExperiment(MontyObjectRecognitionExperiment):
+    def init_loggers(self, logging_config):
+        """Initialize logger with specified log level."""
+        # Add experiment config so config can be passed to wandb
+        all_logging_args = logging_config
+        # all_logging_args.update(config=self.config)
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply dmc/configs/fig9_structured_object_representations.py:82]: yup, done

File: dmc/configs/fig9_structured_object_representations.py (hunk 2 above)
Comment (scottcanoe) [reply dmc/configs/fig9_structured_object_representations.py:215]: Thanks yeah, good catch.

File: dmc/configs/fig8_multimodal_transfer.py (hunk 1 above)
Comment (scottcanoe) [reply dmc/configs/fig8_multimodal_transfer.py:102]: Thanks! I missed that. Corrected.

File: dmc/configs/visualizations.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from copy import deepcopy
+
+from tbp.monty.frameworks.loggers.monty_handlers import DetailedJSONHandler
+
+from .common import DMC_ROOT_DIR
+from .fig4_rapid_inference_with_voting import dist_agent_8lm_half_lms_match
+
+VISUALIZATIONS_DIR = os.path.join(DMC_ROOT_DIR, "visualizations")
+
+config = deepcopy(dist_agent_8lm_half_lms_match)
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply dmc/configs/visualizations.py:11]: Good point. I've added in documentation for this module.

File: monty_capabilities_analysis/scripts/one_lm_summaries.py [hunk 5]
```diff
@@ -189,201 +189,149 @@ def init_1lm_plot(
     return fig
 
 
-def plot_1lm_distant_agent(save: bool = False):
+# def plot_multimodal_transfer_base(save: bool = False):
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply monty_capabilities_analysis/scripts/one_lm_summaries.py:192]: It's been removed.

File: dmc/configs/fig9_structured_object_representations.py [hunk 6]
```diff
@@ -20,28 +20,201 @@
...
+
+Actual behavior:
+- When `MontyExperiment.init_loggers` sees we have a `DetailedJSONHandler`, it 
+  infers that the actual `monty_log_level` is `"DETAILED"`, and uses 
+  `DetailedGraphMatchingLogger` instead of `SelectiveEvidenceLogger`.
+
+To get around this, the following class overrides `init_loggers` which considers
+does everything a detailed log level would require (i.e., telling sensor modules and
+learning modules to store detailed data) without upgrading to "DETAILED". This patch
+could be applied to tbp.monty.
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply dmc/configs/fig9_structured_object_representations.py]: Just commenting here in case it wasn't clear, the "hack" of creating a new experiment subclass for this file is intended to be a short-term solution until the equivalent change has been made to `tbp.monty`.

BTW, this was the issue that got me brainstorming about changes we might want to make to our logging system.

**Comment by scottcanoe** [comment]
Thanks again @nielsleadholm for the feedback! Catching the model path issue for the multimodal transfer experiment was critical. I made that an other requested changes, so I think this PR is ready to go.

On a more general note, I should say that I believe @hlee9212 and I are going to use the `scripts` directory for making figures, and anything there should be considered pretty much constantly in flux until publication. As such, I expect lots of changes there won't need review until much later. In the future, I'll separate out future PRs that contain changes that need full review (i.e., to configs) from PRs @hlee9212 that just have plotting-only changes which won't require the same level of review.

File: dmc/configs/visualizations.py [hunk 7]
```diff
@@ -1,3 +1,22 @@
...
+#
+# Use of this source code is governed by the MIT
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+"""Configs for visualizations (not core experiments)
+
+This file contains configs defined solely for making visualizations that go into
+paper figures. The configs defined are:
+
+- `visualize_8lm_patches`: An experiment that used to save patch view of an object
```
Comment (nielsleadholm) [inline dmc/configs/visualizations.py:15]: that used --> that is used

**Review by nielsleadholm — COMMENTED** [review]
Thanks for those changes Scott, looks great. Just to say that I think there were 3 comments that are still outstanding, you might need to click "Load More" under "3 Hidden Conversations" to see these (see attached screenshots).

And noted about the plotting code, that makes sense to separate them out, thanks!

Screenshot 2025-02-14 at 10 30 25 AM

Screenshot 2025-02-14 at 10 30 13 AM

File: dmc/configs/pretraining_experiments.py [hunk 8]
```diff
@@ -439,6 +446,17 @@ def make_10distinctobj_variant(template: dict) -> dict:
     ),
 )
 
+# Make 10distinctobj variants
+pretrain_dist_agent_1lm_10distinctobj = make_10distinctobj_variant(
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply dmc/configs/pretraining_experiments.py:450]: Ah, right. So we're definitely using this now so I've just removed the NOTE and TODO and added a bit more documentation.

File: dmc/scripts/data_utils.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    # Add 'epoch' column.
+    rows_per_epoch = n_objects * n_lms
+    assert len(df) % rows_per_epoch == 0
+    n_epochs = int(len(df) / rows_per_epoch)
+    df["epoch"] = np.repeat(np.arange(n_epochs), rows_per_epoch)
+
+    return df
+
+
+def get_percent_correct(df: pd.DataFrame) -> float:
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply dmc/scripts/data_utils.py:77]: That's a good point. Yes, I have been using this function pretty much everywhere so far. I'll play around with different options with changes staying local to an analysis script for the time being and revisit this function if that exploration sticks.

**Comment by scottcanoe** [comment]
> Thanks for those changes Scott, looks great. Just to say that I think there were 3 comments that are still outstanding, you might need to click "Load More" under "3 Hidden Conversations" to see these (see attached screenshots).
> 
Oops! Right you are, sorry about that. I've corrected my negligence.

And good point about the accuracy calculation. Definitely something to play with, and it might help us find ways to emphasize the value of our policies.

**Review by nielsleadholm — APPROVED** [review]
No worries, and sounds great!

File: dmc/configs/fig9_structured_object_representations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vkakerbeck) [reply dmc/configs/fig9_structured_object_representations.py]: Oh good catch. I think the automatic setting of log level here https://github.com/thousandbrainsproject/tbp.monty/blob/bb942fc003ee6f711ea9da06bc285318a8e380db/src/tbp/monty/frameworks/experiments/monty_experiment.py#L359 was added to make it less likely to specify a faulty config (here my original commit for that https://github.com/thousandbrainsproject/nupic.monty/commit/49239f94898ee8534430a8ab83059c6c1c6331db). However, I forgot to think about the SELECTIVE log level. Your fix sounds reasonable. We could also remove my initial change and require the user to set the correct log level (maybe just raising a warning)

File: dmc/configs/visualizations.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
+# Copyright 2025 Thousand Brains Project
+# Copyright 2023 Numenta Inc.
```
Comment (vkakerbeck) [inline dmc/configs/visualizations.py:2]: If you just created this file in 2025, you don't need to include Numenta in the copy right header. If you created it last year, it should say 2024 (https://thousandbrainsproject.readme.io/docs/style-guide#source-code-copyright-and-license-header)

File: dmc/scripts/data_utils.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    # Add 'episode' column.
+    assert len(df) % n_lms == 0
+    n_episodes = int(len(df) / n_lms)
+    df["episode"] = np.repeat(np.arange(n_episodes), n_lms)
+
+    # Add 'epoch' column.
+    rows_per_epoch = n_objects * n_lms
+    assert len(df) % rows_per_epoch == 0
+    n_epochs = int(len(df) / rows_per_epoch)
+    df["epoch"] = np.repeat(np.arange(n_epochs), rows_per_epoch)
```
Comment (vkakerbeck) [inline dmc/scripts/data_utils.py:72]: These columns seem generally useful. Should we just add them in the logging part of Monty?

File: dmc/scripts/data_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (vkakerbeck) [reply dmc/scripts/data_utils.py:77]: I think we should definitely count correct_mlh as correct when we report accuracy. This is what we do in our reported benchmarks as well https://github.com/thousandbrainsproject/tbp.monty/blob/0be25350f910a49ea240b3d050aa55e7bb62b1fc/src/tbp/monty/frameworks/loggers/graph_matching_loggers.py#L316 It's basically saying "Monty, you ran out of time and have to make a decision now" and Monty made the right decision. With the ViT we also don't look at how confident it is in it's classification.

**Review by vkakerbeck — COMMENTED** [review]
I just skimmed most of it (let me know if you want me to do a more thorough review of it all) but left one important comment on the accuracy calculation.
